### PR TITLE
pytest: fuzz: fix test not killing fuzz processes correctly

### PR DESCRIPTION
**Pull request by mina86**
`subprocess.call(..., timeout=...)` will send KILL signal to the
process that has been spawned once the timeout passes but that may
leave its descendants still running.  To cover that case, handle
timeout ourselves by waiting and then killing the entire process tree.

### CODE REVIEW

**Review by Ekleog — APPROVED** [review]
Makes sense, I hadn't thought of `psutil`. An alternative would be to start the subprocess with `start_new_session=True` and then call `killpg(getpgid(proc.pid))` similar to what was done in nayduck#28.

**Comment by mina86** [comment]
`psutil` is what NayDuck was using when I took it over and I never changed it even though I think it is a more fragile method.  A test can escape process tree accidentally by doing a double fork while escaping a process group needs more deliberate action.  Nonetheless, so far we didn’t have issue with killing process tree so for now it’s probably good enough.

**Review by Ekleog — APPROVED** [review]

File: pytest/tests/fuzz.py [hunk 1]
```diff
@@ -2,27 +2,64 @@
...
+import signal
 import subprocess
 import sys
 import typing
 
+import psutil
 
-def run(directory: str, fuzz_target: str) -> int:
+
+def run(directory: str, fuzz_target: str, timeout=typing.Optional[int]) -> int:
```
Comment (mm-near) [inline pytest/tests/fuzz.py:14]: what is the unit for timeout ?

File: pytest/tests/fuzz.py [hunk 2]
```diff
@@ -2,27 +2,64 @@
...
+        except psutil.NoSuchProcess:
+            pass
 
 
 def _get_timeout() -> typing.Optional[int]:
+    """Returns test timeout configured on NayDuck
+
+    When run on NayDuck, returns timeout that NayDuck gives us to execute.  If
+    we run over that time, the test will be reported as timed out.  When not
+    running on NayDuck, returns None denoting no timeout.
```
Comment (mm-near) [inline pytest/tests/fuzz.py:61]: what is the unit of the response?

File: pytest/tests/fuzz.py (hunk 1 above)
Comment (mm-near) [inline pytest/tests/fuzz.py:14]: also mention was it the response

**Review by mm-near — CHANGES_REQUESTED** [review]

**Review by mm-near — APPROVED** [review]

**Pull request merged**